code/modules/telesci/dses.dm

```diff
@@ -0,0 +1,291 @@
+/*Deep Space Exploration System
+	used like a sonar system, you ping it, it returns if it finds something with a ping, else a buzz
+
+	POTENTIAL MODULES
+
+ - Ping resonation locator
+	-Tells the user the direction of the caught turf
+ - Ping resonation logging
+	-Logs the co-ordinates of the caught turf
+ - ping resource resequencing
+	-Pings now cost half
+ - ping long-range listener
+	-Ping range doubled
+ - Auto ping system
+	-DSES ping is now togglable, if on it pings every 1 SECOND
+
+*/
+
+/obj/item/device/dses
+	name = "deep space exploration system"
+	desc = "A GPS with a high-gain radio antenna and broadcaster for locating proximity objects in space, the explorers friend."
+	icon_state = "dses"
+	icon = 'icons/obj/telescience.dmi'
+	var/obj/item/weapon/cell/C = null
+	var/module_limit = 1
+	var/list/module_list = list()
+	var/list/locations = list()
+	var/pulse_range = 25
+	var/pulse_cost = 250
+	var/pulse_between = 3 SECONDS
+	var/last_pulse
+	var/last_direction
+	var/last_distance
+	var/list/positive_locations = list()
+	var/auto_pulse = 0
+
+/obj/item/device/dses/attack_self(var/mob/user as mob)
+	if(C)
+		return menu_open(user)
+	else
+		to_chat(user, "<span class = 'notice'>The screen remains dark.</span>")
+		return
+
+/obj/item/device/dses/attackby(obj/item/weapon/W, mob/user)
+	if(istype(W, /obj/item/weapon/dses_module))
+		install_module(W, user)
+	if(istype(W, /obj/item/weapon/cell))
+		if(C)
+			to_chat(user, "<span class = 'notice'>There is already a cell installed.</span>")
+		else
+			if(user.drop_item(W, src))
+				to_chat(user, "<span class = 'notice'>You install \the [W] into the battery slot.</span>")
+				C = W
+	..()
+
+/obj/item/device/dses/proc/menu_open(var/mob/user as mob)
+	var/obj/item/device/dses/t = ""
+	if(C)
+		if(!module_in_list("utoPNG"))
+			t += "<BR><A href='?src=\ref[src];pulse=1'>Pulse</A> "
+		else
+			t += "<BR><A href='?src=\ref[src];tog_pulse=1'>Toggle Pulse:\[[auto_pulse]\]</A>"
+		t += "<BR>Current cell charge: [C.charge]/[C.maxcharge]"
+		t += "<BR>Current device statistics"
+		t += "<BR>Pulse cost: [pulse_cost]"
+		t += "<BR>Pulse range: [pulse_range]"
+		if(module_in_list("dirGET"))
+			t += "<BR><B>Last logged direction: [last_direction]</B>"
+		if(module_in_list("getDST"))
+			t += "<BR><B>Distance from previous ping: [last_distance]</B>"
+		if(module_in_list("crdLOG") && positive_locations.len)
+			t += "<BR>==Logged successful pings=="
```

Comment:
What is the HTML formatting for having a line exactly? First time doing this sorta stuff